OptimizedWalker should skip conditions if the nested node is skipped

E.g.

When using `OptimizedWalker`, the condition created for the "if" attribute of an invocation should be skipped if the invocation is skipped.

```xml
<exec src="empty-script.xml" if="true" />
```

Otherwise the traversal is broken, E.g. using `ClientCompiler`:

```java
io.helidon.build.archetype.engine.v2.InvocationException: Invocation error: Cannot invoke "io.helidon.build.archetype.engine.v2.ast.Node$Builder.doBuild()" because "builder.then" is null

	at archetype.option(compiler/empty-script/main.xml:26)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.Walker.accept(Walker.java:111)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.Walker.walk(Walker.java:131)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.Walker.walk(Walker.java:66)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.OptimizedWalker.walk(OptimizedWalker.java:70)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.ClientCompiler.compile(ClientCompiler.java:204)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.ClientCompilerTest.testEmptyScript(ClientCompilerTest.java:42)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: java.lang.NullPointerException: Cannot invoke "io.helidon.build.archetype.engine.v2.ast.Node$Builder.doBuild()" because "builder.then" is null
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.ast.Condition.<init>(Condition.java:34)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.ast.Condition$Builder.doBuild(Condition.java:163)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.ast.Condition$Builder.doBuild(Condition.java:117)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.ast.Node$Builder.build(Node.java:487)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.ClientCompiler.postVisitBlock(ClientCompiler.java:170)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.ClientCompiler.postVisitBlock(ClientCompiler.java:42)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.ast.Block.acceptAfter(Block.java:336)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.OptimizedWalker.postVisitAny(OptimizedWalker.java:141)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.OptimizedWalker.postVisitBlock(OptimizedWalker.java:116)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.util.OptimizedWalker.postVisitBlock(OptimizedWalker.java:40)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.Walker$DelegateVisitor.postVisitBlock(Walker.java:271)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.ast.Block.acceptAfter(Block.java:336)
	at io.helidon.build.archetype.engine.v2@4.0.0-SNAPSHOT/io.helidon.build.archetype.engine.v2.Walker.accept(Walker.java:109)
	... 74 more
```

Should be back-ported to 3.x

--- a/archetype/engine-v2/src/main/java/io/helidon/build/archetype/engine/v2/ArchetypeEngineV2.java
+++ b/archetype/engine-v2/src/main/java/io/helidon/build/archetype/engine/v2/ArchetypeEngineV2.java
@@ -58,9 +58,15 @@ public class ArchetypeEngineV2 {
     /**
      * Generate a project.
      *
+     * @param externalValues external values
+     * @param externalDefaults external defaults
+     * @param directorySupplier directory supplier
      * @return output directory
      */
-    public Path generate() {
+    public Path generate(Map<String, String> externalValues,
+                         Map<String, String> externalDefaults,
+                         Function<String, Path> directorySupplier) {
+
         Context context = Context.builder()
                                  .cwd(cwd)
                                  .externalValues(externalValues)
@@ -92,6 +98,15 @@ public class ArchetypeEngineV2 {
         }
 
         return directory;
+    }
+
+    /**
+     * Generate a project.
+     *
+     * @return output directory
+     */
+    public Path generate() {
+        return generate(externalValues, externalDefaults, directorySupplier);
     }
 
     /**

--- a/archetype/engine-v2/src/main/java/io/helidon/build/archetype/engine/v2/util/OptimizedWalker.java
+++ b/archetype/engine-v2/src/main/java/io/helidon/build/archetype/engine/v2/util/OptimizedWalker.java
@@ -25,6 +25,7 @@ import java.util.Set;
 
 import io.helidon.build.archetype.engine.v2.Walker;
 import io.helidon.build.archetype.engine.v2.ast.Block;
+import io.helidon.build.archetype.engine.v2.ast.Condition;
 import io.helidon.build.archetype.engine.v2.ast.DeclaredBlock;
 import io.helidon.build.archetype.engine.v2.ast.Invocation;
 import io.helidon.build.archetype.engine.v2.ast.Node;
@@ -114,6 +115,14 @@ public class OptimizedWalker<T> implements Node.Visitor<Void> {
             default:
                 return postVisitAny(block, arg);
         }
+    }
+
+    @Override
+    public Node.VisitResult visitCondition(Condition condition, Void arg) {
+        if (isSkipped(condition.then())) {
+            return Node.VisitResult.SKIP_SUBTREE;
+        }
+        return condition.accept(visitor, this.visitorArg);
     }
 
     @Override

--- a/archetype/engine-v2/src/test/resources/compiler/empty-script/main.xml
+++ b/archetype/engine-v2/src/test/resources/compiler/empty-script/main.xml
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
 
-    Copyright (c) 2022 Oracle and/or its affiliates.
+    Copyright (c) 2022, 2023 Oracle and/or its affiliates.
 
     Licensed under the Apache License, Version 2.0 (the "License");
     you may not use this file except in compliance with the License.
@@ -24,7 +24,7 @@
         <inputs>
             <enum id="enum1" name="Enum1">
                 <option value="enum1-value1" name="Enum1Value1">
-                    <exec src="script1.xml"/>
+                    <exec src="script1.xml" if="true"/>
                 </option>
                 <option value="enum1-value2" name="Enum1Value2"/>
             </enum>

